code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java

```diff
@@ -277,6 +298,38 @@ public void releaseAllocation(Volume volume) {
     }
 
     @Override
+    public List<Long> getInstancesWithVolumeMounted(long volumeId, long currentInstanceId) {
+        return create()
+                .select(INSTANCE.ID)
+                .from(INSTANCE)
+                .join(MOUNT)
+                    .on(MOUNT.INSTANCE_ID.eq(INSTANCE.ID).and(MOUNT.VOLUME_ID.eq(volumeId)))
+                .join(INSTANCE_HOST_MAP)
+                    .on(INSTANCE_HOST_MAP.INSTANCE_ID.eq(INSTANCE.ID))
+                .where(INSTANCE.REMOVED.isNull()
+                    .and(INSTANCE.ID.ne(currentInstanceId))
+                    .and(INSTANCE_HOST_MAP.STATE.notIn(IHM_STATES))
+                    .and((INSTANCE.HEALTH_STATE.isNull().or(INSTANCE.HEALTH_STATE.eq(HealthcheckConstants.HEALTH_STATE_HEALTHY)))))
```

Comment:
THis is a little weird, but I added the healthstate check so that a volume can be moved if it is currently on a host where the instances that are using it are unhealthy.

This is needed for the basic scenario of a host going down an need to migrate a vm/container to a different host.